### PR TITLE
Remi behind jupyter server proxy

### PR DESCRIPTION
**Pull request by gbrault**
- Designed with minimal remi change in mind (see diff files to estimate the level of change)
- For both remi core and remi apps (minimal change)
- Fully tested with jupyter-server-proxy
- Jupyter notebooks provided
- works only with 3.6+ (f"" string interpolation)
- If proxy not used, legacy remi app are working
- adding proxy feature need some remi app rework (to calculate url for the front-end)

```python
"url('/editor_resources:resize.png')" => f"url({_proxy.set_url('/editor_resources:resize.png')})"
# if directly provided to front-end
# if url are passed to Image or the like, no change
```

### CODE REVIEW

**Comment by dddomodossola** [comment]
@gbrault this is an extraordinary work! The work that can be done only by an experienced man. Please EXCUSE ME for the late reply. 
I see f strings and Proxy a bit invasive. 

Fstrings are fine for performances and readability, but it prevents the use on older python versions. Just to make you an example, I'm using remi on a plc (beckhoff softplc) that runs only python 3.4. Unfortunately this happens. 

Proxy seems a bit invasive, I'm not sure to understand why it is required. Can you please explain it simply to me?

**Comment by gbrault** [comment]
@dddomodossola I understand your point! I am interested by remi as a gui component for Jupyter based applications. I use Jupyter for processing data from production and sales for a medium company and time to time, I need some UI to enter data or browse sql database. Remi seems a good candidate for that.
Of course I understand my goal is not fully aligned with yours for making embedded application.
Maybe the gap implies to have 2 different version for remi?

**Comment by gbrault** [comment]
Of course, I need a proxy to have both, the notebook and the UI interoperating seemlessly. The user see it as a single application. Starting the UI from a notebook enable to share the same python machine.

**Comment by dddomodossola** [comment]
> @dddomodossola I understand your point! I am interested by remi as a gui component for Jupyter based applications. I use Jupyter for processing data from production and sales for a medium company and time to time, I need some UI to enter data or browse sql database. Remi seems a good candidate for that. Of course I understand my goal is not fully aligned with yours for making embedded application. Maybe the gap implies to have 2 different version for remi?

I'm really interested in your changes, lots of people uses jupyter and of course it is wonderful to get remi usable from it. Maybe this is feasible without affecting remi server. I have to study your changes in detail.

> Of course, I need a proxy to have both, the notebook and the UI interoperating seemlessly. The user see it as a single application. Starting the UI from a notebook enable to share the same python machine.

Maybe the Proxy class can be avoided (not sure). I'm thinking about solving the resource problem directly in remi server:
1. We can leave all the urls as is, i.e. "/res:myicon.png"
2. At application runtime, the resource request arrives to the server as a GET request. Normally the server loads and sends the resource content to the client. In case of a jupyter app, the server **_get_static_file** function gets overridden, to load/redirect the request to jupyter. The function to override is just **App._get_static_file**.

If this solution could be feasible, the remi library could get untouched, and all gets solved by a template App, with a special  **_get_static_file** function.

```python
import remi.gui as gui
from remi import start, App
import os

class MyApp(App):
    def _get_static_file(self, filename):
        #here the magic should be done
        ...

    def idle(self):
        ...

    def main(self):
        ...


if __name__ == "__main__":
    start(MyApp)
```

Do you think this could be feasible? Thank you a lot for the collaboration.

**Comment by gbrault** [comment]
This is a good idea!

```python
import remi.gui as gui
from remi import start, App
import os

class MyApp(App):
    def _get_static_file(self, filename):
        #here the magic should be done
        ...

    def idle(self):
        ...

    def main(self):
        ...


if __name__ == "__main__":
    start(MyApp)
```

This solve all the cases where the client ask for a file and in _get_static_file, we have a chance to tweak all content where you have patterns like "/<res>:" where res can be in ['res', "editor_resource', etc...]

I can give a try to write "#here the magic should be done"... (it should be mainly based on some code I wrote already)

I let you know.

**Comment by gbrault** [comment]
@dddomodossola, I used your recommendation, it comes to only modifying server.py

1. adding _net_interface_ip function to App class to be able to overload it (this is for the ws(s) address)
2. creating an _overload function to App class to be able to overload it (to replace /res: with /proxy/8085/res:, not restricted to res)
3. adding **kwards in the prototyp of __process_all
4. Performing a call to _overload in every line of code (of App) which return html content to  (to replace /res: with /proxy/8085/res:, not restricted to res)

I don't think I have used coding rules which cannot be used in earlier version of Python in server.py.

I calling code, it can be post 3.6+ version (no influence on your code) 

If you want we can make a video call to show you it works and see if we can even simplify again!

Let me know!

**Comment by gbrault** [comment]
I have also created all_paths to get all the /res: like tokens

**Comment by gbrault** [comment]
One drawback of this method might be performance:
- instead of calculating the right url (like in the Proxy version), it is done on the whole content for each request
- this can be "time consuming"  on small targets
- however, if the replacement is not mad, we only add one call per content request, which is not much

**Comment by gbrault** [comment]
Using IFrame, it works in a Notebook cell!
image

**Comment by gbrault** [comment]
image

**Comment by dddomodossola** [comment]
@gbrault you did the magic! I gave a rapid look at the code and seems:
1. Readable
2. Simple
3. Non invasive

I will analyze it better this late evening and merge. Thank you so much!

File: remi/server.py [hunk 1]
```diff
@@ -631,18 +651,21 @@ def _get_static_file(self, filename):
...
         return os.path.join(paths[key], path)
+    
+    def _overload(self, data, **kwargs):
+        """Used to overload the content before sent back to client"""
+        return data
 
-    def _process_all(self, func):
+    def _process_all(self, func, **kwargs):
         self._log.debug('get: %s' % func)
+        if 'overload' not in kwargs:
```
Comment (dddomodossola) [inline remi/server.py:665]: Why is it required to pass the overload function as arguments? The App class is inherited, so the function **_overload** is already overloaded. Doesn't it?

**Comment by dddomodossola** [comment]
Why you preferred to replace/correct paths instead of resolving them by overloading the **_get_static_file** function? I supposed that overloading that _get_static_file function you can interpret the path inside it. Isn't this possible?

**Comment by gbrault** [comment]
From the reverse proxy perspective (Jupyter-Server-Proxy), an url targeting the remi server, in the client (browser) needs to be
```python
http(s)://<the server domain>:{jlabport}/proxy/{remiport}/<original path to remi server>
```
When remi generates url, the /proxy/{remiport}/ does not exists and it would rather send to the client
```python
http(s)://<the server domain>:{remiport}/<original path to remi server>
```

From my understanding _get_static_file function would be usefull to find the os disk path if given
```python
/proxy/{remiport}/<original path to remi server>
```
But this never happens, as the reverse proxy when called by the client with
```python
http(s)://<the server domain>:{jlabport}/proxy/{remiport}/<original path to remi server>
```
calls back remi server with
```python
http://localhost:{remiport}/<original path to remi server>
```

Do you agree, or am I missing something?

**Comment by dddomodossola** [comment]
Yes, thank you for the clarification. 
Can you please also explain the code comment above?

File: remi/server.py (hunk 1 above)
Comment (gbrault) [reply remi/server.py:665]: You are perfectly right, this is not needed, we just need to use self._overload to call the function (it is a scorie which need to be cleaned-up: I make  the correction and send new a new version.

**Comment by dddomodossola** [comment]
Thank you ;-)

**Comment by gbrault** [comment]
The extra code in _process_all is cleaned

**Comment by dddomodossola** [comment]
Merged, thank you 😊

**Comment by gbrault** [comment]
You welcome!

**Comment by dddomodossola** [comment]
Now remi can be executed from Jupyter, this is cool

**Comment by gbrault** [comment]
With this feature, when you have target platform able to support Jupyter Lab (I used it on a raspberry PI 3 A+ with success), you have a complete web access with:
- remi as the UI 
- Notebooks
- Jlab as the IDE (including a Terminal)

**Pull request merged**